# Лабораторная работа 2 
## Выполнил студент группы БВТ1905 Бернадский Илья Дмитриевич
### Задание 
#### Реализовать методы поиска в соответствии с заданием. Организовать генерацию начального набора случайных данных. Для всех вариантов добавить реализацию добавления, поиска и удаления элементов. Оценить время работы каждого алгоритма поиска и сравнить его со временем работы стандартной функции поиска, используемой в выбранном языке программирования.

### Генерация массива
#### (Создал два массива - один побольше и второй поменьше для тестов)

In [73]:
import time
from random import randint
big_mass = []
lil_mass = []
tree_mass = []
for i in range(10000000):
    big_mass.append(randint(-10000, 10000))
for i in range(100000):
    lil_mass.append(randint(-10000, 10000))
for i in range(100000):
    tree_mass.append(randint(-10000000, 10000000))
lil_mass.append(3456)
tree_mass.append(3456)
big_mass.sort()
lil_mass.sort()

### Функция для тестов

In [2]:
def test(found_big,found_lil):
    global sought 
    global big_mass 
    global lil_mass
    global start_time
    print("Искомый элемент тут: ", found_big)
    print("Затраченное время на поиск алгоритмом ", time.time() - start_time)
    start_time = time.time()
    a = big_mass.index(sought)
    print("Затраченное время на поиск стандартной функцией ", time.time() - start_time)

    start_time = time.time()
    print("Искомый элемент тут: ", found_lil)
    print("Затраченное время на поиск алгоритмом ", time.time() - start_time)
    start_time = time.time()
    a = lil_mass.index(sought)
    print("Затраченное время на поиск стандартной функцией ", time.time() - start_time)

## Задача 1

### Бинарный поиск

#### В упорядоченном массиве находится средний элемент. Дальше 3 варианта развития событий:
#### 1. Если средний элемент равен искомому, поиск заканчивается
#### 2. Если средний элемент больше искомого, то правая граница сдвигается к среднему и опять находится средний
#### 3. Если средний элемент меньше искомого, то левая граница сдвигается к среднему и опять находится средний
#### Повторяется пока не найдет

In [3]:
def binary_search(mass, sought):
    left = 0
    right = len(mass) - 1
    mid = (left + right) // 2
    while mass[mid] != sought and left <= right:
        if sought > mass[mid]:
            left = mid + 1
        else:
            right = mid - 1
        mid = (left + right) // 2

    if left > right :
        return "Нету"
    else:
        return mid
        

In [4]:
sought = int(input("Введите искомое число "))
start_time = time.time()
test(binary_search(big_mass, sought), binary_search(lil_mass, sought))

Введите искомое число 3456
Искомый элемент тут:  6729124
Затраченное время на поиск алгоритмом  0.0
Затраченное время на поиск стандартной функцией  0.38735270500183105
Искомый элемент тут:  67296
Затраченное время на поиск алгоритмом  0.0
Затраченное время на поиск стандартной функцией  0.0020017623901367188


#### Можно сделать вывод, что бинарный поиск ищет элемент быстрее чем стандартная функция

### Фибоначчиев поиск
#### Элемент ищется по индексам равным числам фибоначчи 
#### Если очередной элемент больше искомого, то массив обрезается так, что левая граница это прошлое число фибоначчи, а правая это последнее взятое. Повторять пока очередной элемент не будет равен искомому

In [9]:
def fibonacci_search(mass, sought):
    obrez = 0
    past = 1
    present = 1
    while mass[past] != sought:
        past = 1
        present = 1
        while mass[present - 1] < sought:
            bubble = present
            present += past
            past = bubble
            if present > len(mass):
                present = len(mass)
        mass = mass[:present]
        mass = mass[past:]
        obrez += past 
        present = len(mass)
        past = 0
    return obrez

In [10]:
sought = int(input("Введите искомое число "))
start_time = time.time()
big_mass_copy = big_mass.copy()
lil_mass_copy = lil_mass.copy()
test(fibonacci_search(big_mass_copy, sought), fibonacci_search(lil_mass_copy, sought))

Введите искомое число 3456
Искомый элемент тут:  6728858
Затраченное время на поиск алгоритмом  0.5284817218780518
Затраченное время на поиск стандартной функцией  0.39235687255859375
Искомый элемент тут:  67296
Затраченное время на поиск алгоритмом  0.0
Затраченное время на поиск стандартной функцией  0.0020017623901367188


#### Фибоначчиев поиск ищет быстрее в маленьких массивах, а стандартная функция быстрее в больших

### Интерполяционный поиск
#### Схож с бинарным поиском, но средний элемент высчитывается не средним значением, а специальной формулой

In [11]:
def interpolation_search(mass, sought):
    left = 0
    right = (len(mass) - 1)
    while left <= right and sought >= mass[left] and sought <= mass[right]:
        index = left + int(((float(right - left) / ( mass[right] - mass[left])) * ( sought - mass[left])))
        if mass[index] == sought:
            return index
        if mass[index] < sought:
            left = index + 1
        else:
            right = index - 1

In [12]:
sought = int(input("Введите искомое число "))
start_time = time.time()
big_mass_copy = big_mass.copy()
lil_mass_copy = lil_mass.copy()
test(interpolation_search(big_mass_copy, sought), interpolation_search(lil_mass_copy, sought))

Введите искомое число 3456
Искомый элемент тут:  6728999
Затраченное время на поиск алгоритмом  0.27324795722961426
Затраченное время на поиск стандартной функцией  0.37598109245300293
Искомый элемент тут:  67300
Затраченное время на поиск алгоритмом  0.0
Затраченное время на поиск стандартной функцией  0.0010056495666503906


#### Интерполяционный поиск немного быстрее стандартной функции

### Бинарное дерево
#### Строится бинарное дерево, в котором у каждого элемента может быть две ветки: левая с объектом меньше и правая с объектом больше. Поиск по этому дереву ведется по принципу - если искомое число меньше, чем сравниваемое число в дереве, то сравнение продолжается по левой ветке сравниваемого числа, если больше, то по правой. Перебирание веток происходит пока не найдется искомое число

In [80]:
class Tree:

    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data

    def insert(self, data):
        if self.data:
            if data < self.data:
                if self.left is None:
                    self.left = Tree(data)
                else:
                    self.left.insert(data)
            elif data > self.data:
                if self.right is None:
                    self.right = Tree(data)
                else:
                    self.right.insert(data)
        else:
            self.data = data

    def findNum(self, lkpval):
        if lkpval < self.data:
            if self.left is None:
                return str(lkpval)+" Not Found"
            return self.left.findNum(lkpval)
        elif lkpval > self.data:
            if self.right is None:
                return str(lkpval)+" Not Found"
            return self.right.findNum(lkpval)
        else:
            print(str(self.data) + ' is found')

    def PrintTree(self):
        if self.left:
            self.left.PrintTree()
        print( self.data),
        if self.right:
            self.right.PrintTree()

#### Создание дерева по массиву

In [81]:
tree = Tree(tree_mass[0])
for i in range(1, len(tree_mass)):
    tree.insert(tree_mass[i])

#### Поиск по дереву

In [82]:
start_time = time.time()
tree.findNum(3456)
print(time.time() - start_time)

3456 is found
0.0010004043579101562


### Все результаты

#### 1. 
#### Бинарное дерево (по маленькому массиву) 0.0010004043579101562 секунд
#### 2. 
#### Бинарный поиск (по большому массиву) 0.0 секунд (слишком быстро)
#### Бинарный поиск (по маленькому массиву) 0.0 секунд (слишком быстро)
#### 3. 
#### Фибоначчиев (по большому массиву) 0.5284817218780518 секунд
#### Фибоначчиев (по маленькому массиву) 0.0 секунд (слишком быстро)
#### 4. 
#### Интерполяционный (по большому массиву) 0.27324795722961426 секунд
#### Интерполяционный (по маленькому массиву) 0.0 секунд (слишком быстро)
#### 5. 
#### Стандартная функция (по большому массиву) 0.37598109245300293 секунд
#### Стандартная функция (по маленькому массиву) 0.0010056495666503906 секунд

## Задача 2
#### Функция хеширования 

In [5]:
def hash(number, size):
    number += 10000
    size = size // 20000
    return number * size

#### Массив для теста рехеширования

In [3]:
hash_mass = []
for i in range(30000):
    hash_mass.append(randint(-10000, 10000))

### Простое рехеширование
#### Если место для очередного элемента занято, то рассматривается следующее место, затем еще следующее и т.д.

In [43]:
start_time = time.time()
rehashed = [None] *len(hash_mass)
size = len(hash_mass)
for j in range(len(hash_mass)):
    i = 0
    if rehashed[hash(hash_mass[j],size)] == None:
        rehashed[hash(hash_mass[j], size)] = hash_mass[j]
    else:
        while rehashed[hash(hash_mass[j], size) + i % size] != None:
            i += 1
        rehashed[hash(hash_mass[j], size) + i % size] = hash_mass[j]
simpleHashTime = time.time() - start_time
print(simpleHashTime)

101.68108129501343


### Рехеширование с помощью псевдослучайных чисел
#### Если место для очередного элемента занято, для него выбирается псевдослучайное место, до тех пор, пока не найдется подходящее.

In [65]:
start_time = time.time()
def randHash(number, size):
    return hash(number, size)
rehashed = [None] *len(hash_mass)
size = len(hash_mass)
for j in range(len(hash_mass)):
    i = 0
    if rehashed[randHash(hash_mass[j],size)] == None:
        rehashed[randHash(hash_mass[j], size)] = hash_mass[j]
    else:
        while rehashed[i] != None:
            i = randint(1, size - 1)
        rehashed[i] = hash_mass[j]
randomHashTime = time.time() - start_time
print(randomHashTime)

0.5194735527038574


### Рехеширование цепочкой
#### Если место для очередного элемента занято, создается цепочка, в которую будут добавлятся все элементы с таким же кодом

In [64]:
start_time = time.time()
rehashed = [None] * len(hash_mass)
for i in range(len(hash_mass)):
    if rehashed[hash(hash_mass[i], len(hash_mass))] == None:
        rehashed[hash(hash_mass[i], len(hash_mass))] = hash_mass[i]
    elif rehashed[hash(hash_mass[i], len(hash_mass))] != None:
        chain = rehashed[hash(hash_mass[i], len(hash_mass))]
        if type(chain) != "list":
            rehashed[hash(hash_mass[i], len(hash_mass))] = []
            if type(chain) is list:
                for j in range(len(chain)):
                    rehashed[hash(hash_mass[i], len(hash_mass))].append(chain[j])
            else:
                rehashed[hash(hash_mass[i], len(hash_mass))].append(chain)
        rehashed[hash(hash_mass[i], len(hash_mass))].append(hash_mass[i])
chainHashTime = time.time() - start_time
print(chainHashTime)

0.08507657051086426


### Сравнение всех результатов

In [66]:
print("Простое рехеширование", simpleHashTime)
print("Рехеширование рандомом", randomHashTime)
print("Рехеширование цепочками", chainHashTime)

Простое рехеширование 101.68108129501343
Рехеширование рандомом 0.5194735527038574
Рехеширование цепочками 0.08507657051086426


## Задача 3

### Создадим доску в виде матрицы

In [1]:
import numpy as np
def output(matrix):
    max_len = max([len(str(e)) for r in matrix for e in r])
    for row in matrix:
        print(*list(map('{{:>{length}}}'.format(length=max_len).format, row)))
matrix = np.array([[0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0], ])
output(matrix)

0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0


### Сделаем две функции:
#### 1. Вызывается когда находится свободное место для ферзя, ставит его, вызывает вторую функцию, и ищет свободное место в следующем ряду, после чего вызывает эту же функцию. Выполняется пока не найдутся все варианты
#### 2. Заполняет двойками клетки слева, справа, сверху, снизу и по диагонали от поставленного только что ферзя, чтобы обозначить клетки, на которые ферзя ставить нельзя

In [2]:
def zapolnyaem(x, y, doska):
    for i in range(8):
        for j in range(8):
            if i == x or j == y or abs(i - x) == abs(j - y):
                doska[i][j] = 2
    return doska
def oppaPlusFerz(x, y, matrica, k):
    global variant
    kolvo = k + 1
    doska = matrica.copy()
    doska = zapolnyaem(x, y, doska)
    doska[x][y] = 1
    for i in range(x + 1, 8):
        for j in range(8):
            if doska[i][j] == 0:
                oppaPlusFerz(i, j, doska, kolvo)
            elif j == 7:
                return
    if kolvo == 8:
        variant += 1
        print(variant)
        output(doska)
        print()
variant = 0
for i in range(8):
    oppaPlusFerz(0, i, matrix, 0)

1
1 2 2 2 2 2 2 2
2 2 2 2 1 2 2 2
2 2 2 2 2 2 2 1
2 2 2 2 2 1 2 2
2 2 1 2 2 2 2 2
2 2 2 2 2 2 1 2
2 1 2 2 2 2 2 2
2 2 2 1 2 2 2 2

2
1 2 2 2 2 2 2 2
2 2 2 2 2 1 2 2
2 2 2 2 2 2 2 1
2 2 1 2 2 2 2 2
2 2 2 2 2 2 1 2
2 2 2 1 2 2 2 2
2 1 2 2 2 2 2 2
2 2 2 2 1 2 2 2

3
1 2 2 2 2 2 2 2
2 2 2 2 2 2 1 2
2 2 2 1 2 2 2 2
2 2 2 2 2 1 2 2
2 2 2 2 2 2 2 1
2 1 2 2 2 2 2 2
2 2 2 2 1 2 2 2
2 2 1 2 2 2 2 2

4
1 2 2 2 2 2 2 2
2 2 2 2 2 2 1 2
2 2 2 2 1 2 2 2
2 2 2 2 2 2 2 1
2 1 2 2 2 2 2 2
2 2 2 1 2 2 2 2
2 2 2 2 2 1 2 2
2 2 1 2 2 2 2 2

5
2 1 2 2 2 2 2 2
2 2 2 1 2 2 2 2
2 2 2 2 2 1 2 2
2 2 2 2 2 2 2 1
2 2 1 2 2 2 2 2
1 2 2 2 2 2 2 2
2 2 2 2 2 2 1 2
2 2 2 2 1 2 2 2

6
2 1 2 2 2 2 2 2
2 2 2 2 1 2 2 2
2 2 2 2 2 2 1 2
1 2 2 2 2 2 2 2
2 2 1 2 2 2 2 2
2 2 2 2 2 2 2 1
2 2 2 2 2 1 2 2
2 2 2 1 2 2 2 2

7
2 1 2 2 2 2 2 2
2 2 2 2 1 2 2 2
2 2 2 2 2 2 1 2
2 2 2 1 2 2 2 2
1 2 2 2 2 2 2 2
2 2 2 2 2 2 2 1
2 2 2 2 2 1 2 2
2 2 1 2 2 2 2 2

8
2 1 2 2 2 2 2 2
2 2 2 2 2 1 2 2
1 2 2 2 2 2 2 2
2 2 2 2 2 2 1 2
2 2 2 1 2 2 2 2
2

82
2 2 2 2 2 2 1 2
2 1 2 2 2 2 2 2
2 2 2 1 2 2 2 2
1 2 2 2 2 2 2 2
2 2 2 2 2 2 2 1
2 2 2 2 1 2 2 2
2 2 1 2 2 2 2 2
2 2 2 2 2 1 2 2

83
2 2 2 2 2 2 1 2
2 1 2 2 2 2 2 2
2 2 2 2 2 1 2 2
2 2 1 2 2 2 2 2
1 2 2 2 2 2 2 2
2 2 2 1 2 2 2 2
2 2 2 2 2 2 2 1
2 2 2 2 1 2 2 2

84
2 2 2 2 2 2 1 2
2 2 1 2 2 2 2 2
1 2 2 2 2 2 2 2
2 2 2 2 2 1 2 2
2 2 2 2 2 2 2 1
2 2 2 2 1 2 2 2
2 1 2 2 2 2 2 2
2 2 2 1 2 2 2 2

85
2 2 2 2 2 2 1 2
2 2 1 2 2 2 2 2
2 2 2 2 2 2 2 1
2 1 2 2 2 2 2 2
2 2 2 2 1 2 2 2
1 2 2 2 2 2 2 2
2 2 2 2 2 1 2 2
2 2 2 1 2 2 2 2

86
2 2 2 2 2 2 1 2
2 2 2 1 2 2 2 2
2 1 2 2 2 2 2 2
2 2 2 2 1 2 2 2
2 2 2 2 2 2 2 1
1 2 2 2 2 2 2 2
2 2 1 2 2 2 2 2
2 2 2 2 2 1 2 2

87
2 2 2 2 2 2 1 2
2 2 2 1 2 2 2 2
2 1 2 2 2 2 2 2
2 2 2 2 2 2 2 1
2 2 2 2 2 1 2 2
1 2 2 2 2 2 2 2
2 2 1 2 2 2 2 2
2 2 2 2 1 2 2 2

88
2 2 2 2 2 2 1 2
2 2 2 2 1 2 2 2
2 2 1 2 2 2 2 2
1 2 2 2 2 2 2 2
2 2 2 2 2 1 2 2
2 2 2 2 2 2 2 1
2 1 2 2 2 2 2 2
2 2 2 1 2 2 2 2

89
2 2 2 2 2 2 2 1
2 1 2 2 2 2 2 2
2 2 2 1 2 2 2 2
1 2 2 2 2 2 2 2
2 2 2 2 2